### PR TITLE
 Fixing race condition for early parcels

### PR DESCRIPTION
**Pull request by sithhell**
This patch fixes a race condition between creating the scheduler queues
in the thread pool and setting its state to running.
In addition, we need to wait until the thread pool is running before scheduling
tasks onto it.

Flyby: relaxing lock ignoring for condition_variable

### CODE REVIEW

File: hpx/runtime/applier/apply_helper.hpp [hunk 1]
```diff
@@ -80,6 +81,12 @@ namespace hpx { namespace applier { namespace detail
                 static_cast<threads::thread_stacksize>(
                     traits::action_stacksize<Action>::value));
 
+            while (!threads::threadmanager_is_at_least(state_running))
+            {
+                boost::this_thread::sleep(boost::get_system_time() +
```
Comment (biddisco) [inline hpx/runtime/applier/apply_helper.hpp]: I presume that the use of a sleep function is a case of last resort and there is no other way of waiting until the runtime is ready.

File: src/state.cpp [hunk 2]
```diff
@@ -26,6 +26,15 @@ namespace hpx
...
             return (rt->get_thread_manager().status() == st);
         }
+        bool threadmanager_is_at_least(state st)
+        {
+            hpx::runtime* rt = get_runtime_ptr();
+            if (nullptr == rt) {
+                // we're probably either starting or stopping
+                return false;
+            }
+            return (rt->get_thread_manager().status() >= st);
```
Comment (biddisco) [inline src/state.cpp]: I hope that the states always form a progression tha increases with some understood metric of what a threadmanager should be doing.

File: hpx/runtime/applier/apply_helper.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/runtime/applier/apply_helper.hpp]: Correct. There is not much we can do otherwise.

File: src/state.cpp (hunk 2 above)
Comment (hkaiser) [inline src/state.cpp]: They are, indeed.

**Pull request merged**